feat(search): Add support for `is:linked` and `is:unlinked` filters

saifelse opened the pull request:
- "linked" means there is an issue linked to the group.
- "unlinked" means there isn't an issue linked to the group.

The implementation mirrors that of the `is:assigned`/`is:unassigned`
filters where it does a non-snuba, postgres filter for `Group`s
that are either linked to:
    - an `ExternalIssue` (via `GroupLink` with `LinkedType.issue`), or
    - a `PlatformExternalIssue`

Fixes #15725, fixes #8338, fixes #5413, fixes #15525.

## Screenshots
These are outdated as the keywords were changed to `linked` and `unlinked`

### Issue search
![image](https://user-images.githubusercontent.com/549473/97155023-f23f4080-1731-11eb-8a9b-24d77193da6e.png)
![image](https://user-images.githubusercontent.com/549473/97154045-ac35ad00-1730-11eb-99a1-b8b0fa9b81f1.png)
![image](https://user-images.githubusercontent.com/549473/97154087-b9529c00-1730-11eb-8acb-e0b21475ee2f.png)

saifelse commented on src/sentry/search/utils.py:
```diff
@@ -427,6 +427,10 @@ def parse_query(projects, query, user, environments):
                     results["unassigned"] = True
                 elif value == "assigned":
                     results["unassigned"] = False
+                elif value == "triaged":
+                    results["triaged"] = True
+                elif value == "untriaged":
+                    results["triaged"] = False
```
I believe this is only used for event search. By specifying these we get the desired error message:
![image](https://user-images.githubusercontent.com/549473/97155225-41857100-1732-11eb-8049-12602b85ac9e.png)
(instead of it complaining about unknown "is" attributes)

iProgramStuff commented on src/sentry/search/snuba/backend.py:
```diff
@@ -49,6 +50,19 @@ def unassigned_filter(unassigned, projects):
     return query
 
 
+def triaged_filter(triaged, projects):
+    query = Q(
+        id__in=GroupLink.objects.filter(
+            project_id__in=[p.id for p in projects],
+            linked_type=GroupLink.LinkedType.issue,
+            relationship=GroupLink.Relationship.references,
+        ).values_list("group_id", flat=True)
+    )
+    if not triaged:
+        query = ~query
+    return query
```
This looks good to me, but will miss groups linked via our integration platform (https://docs.sentry.io/product/integrations/integration-platform/). I think we'd need an OR here on the  PlatformExternalIssue model.

iProgramStuff commented:
Nice PR. Relaying some information I gathered from the team here (thanks @markstory!). Apparently, Mark actually built this feature awhile back but didn't ship it because we thought this was going to be handled within Snuba/Clickhouse as we brought more group data into there....but that hasn't happened _yet_. His closed PR is here: https://github.com/getsentry/sentry/pull/13854

The solution in general looks good, other than missing some of the links due to the PlatformExternalIssue issue.

saifelse commented on src/sentry/static/sentry/app/stores/tagStore.tsx:
```diff
@@ -86,6 +86,8 @@ const tagStoreConfig: Reflux.StoreDefinition & TagStoreInterface = {
           'muted',
           'assigned',
           'unassigned',
+          'triaged',
+          'untriaged',
```
These aren't immediately surfaced:
![image](https://user-images.githubusercontent.com/549473/97192702-62b38500-1765-11eb-961e-257e5987319d.png)

until you type enough to filter out the earlier options:
![image](https://user-images.githubusercontent.com/549473/97192722-6a732980-1765-11eb-8643-f55930e94837.png)

If we use `is:` for this / other orthogonal states, I think it's important for the UI to always enumerate all the options for feature discoverability.

Nice to have: a tooltip / description that explains what each option means... maybe the other options are more self-explanatory... but I worry triaged/untriaged might not be obvious.

saifelse replied:
Handling in #21719

MeredithAnya commented:
hey @saifelse ! Great PRs you've been creating on Sentry ! I'm on the Ecosystem team here at Sentry, and just wanted to give my thoughts here re: naming. To me `triage` and `untriaged` are too vague of terms to be used to denote that the Sentry issue is linked to an external service. 

As you said:
> but I worry triaged/untriaged might not be obvious.

and I agree. I see a linked issue being part of the triage process but I don't see it as the only thing (assigning a member/team could be part of the process for example). Even though we do have assigned/unassigned already for that, I still think triage/untriaged could be a bit confusing for this case.

I could see something like `is:tracked` or `is:untracked` as an option, but I'm not sure I'm sold on that either. `linked_ticket:True` is certainly more explicit.

BYK commented on src/sentry/search/snuba/backend.py:
```diff
@@ -49,6 +50,19 @@ def unassigned_filter(unassigned, projects):
+            project_id__in=[p.id for p in projects],
+            linked_type=GroupLink.LinkedType.issue,
+            relationship=GroupLink.Relationship.references,
+        ).values_list("group_id", flat=True)
```
Don't know how our search works but it _feels_ to me that using `.exists()` here would be more appropriate and efficient.

saifelse replied:
I believe this is building onto a query that already does `SELECT ... FROM group`  and this is an additional filter for groups that is tacked on. `SELECT * FROM group [[WHERE id IN (...)]]`, where this function is basically providing an additional WHERE clause.

Using `exists()` to me in some capacity would imply we're doing a correlated subquery, .e.g`WHERE EXISTS (SELECT group_link WHERE group_id=group.id)`, which seems undesirable. I think theoretically doing a JOIN to group_link for link-existence and an OUTER JOIN for non-existence can work and allows the postgres planner to be more efficient, but in my experience that doesn't work well with query builders like this since it requires coordination between this filter and other filters (e.g. making sure the joins are aliased, or re-using pre-existing joins in the query).

FWIW, I'm pattern matching off of how assignees are built, so it should be roughly as efficient as that.

saifelse commented:
@MeredithAnya:

> hey @saifelse ! Great PRs you've been creating on Sentry !

Thanks! It's been fun digging into this codebase and feeling empowered to implement things that I've been yearning for 😇 .

> As you said:
> 
> > but I worry triaged/untriaged might not be obvious.
> 
> and I agree. I see a linked issue being part of the triage process but I don't see it as the only thing (assigning a member/team could be part of the process for example). Even though we do have assigned/unassigned already for that, I still think triage/untriaged could be a bit confusing for this case.
> 
> I could see something like `is:tracked` or `is:untracked` as an option, but I'm not sure I'm sold on that either. `linked_ticket:True` is certainly more explicit.

If we stick to the the terminology alreayd used in the UI, "linked issue" seems most appropriate (just "issue" is probably too ambiguous since Sentry Issues exist).

I don't know if it'll be technically complex yet, but treating it like a tag might be the most intuitive to a user and is a bit more flexible, allowing the query syntax `has:linked-issued` / `linked-issue:APP-16`? It can still be fairly discoverable since it can now be present as its own dedicated field in the Search Builder UI. Thoughts?

wedamija commented:
Hi @saifelse, just want to join the others in thanking you for your pr!

We've been discussing the naming internally, and after a lot of 🚲 🏠  we think that `is:linked` will probably fit best. We'd prefer to not treat it like a tag, since that then means whichever name we choose becomes reserved, and might conflict with how existing users have their data set up already.

I think the general approach here looks good, and think we could merge this once we include `PlatformExternalIssue` as part of the search, and get some test coverage in place.

saifelse committed:
feat(search): Add support for `is:linked` and `is:unlinked` filters

- "linked" means there is an issue linked to the group.
- "unlinked" means there isn't an issue linked to the group.

The implementation mirrors that of the `is:assigned`/`is:unassigned`
filters where it does a non-snuba, postgres filter for `Group`s
that are either linked to:
    - an `ExternalIssue` (via `GroupLink` with `LinkedType.issue`), or
    - a `PlatformExternalIssue`

Fixes #15725, fixes #8338, fixes #5413, fixes #15525.

saifelse commented on src/sentry/search/snuba/backend.py:
```diff
@@ -49,6 +51,27 @@ def unassigned_filter(unassigned, projects):
 
 
+def linked_filter(linked, projects):
+    # Note that there is no filter being applied here. We rely on the database's query optimizer to
+    # recognize `WHERE id IN (SELECT group_id FROM sentry_platformexternalissue)` the same as
+    # `JOIN sentry_platformexternalissue ON sentry_platformexternalissue.group_id = id` and then choose
+    # the appropriate indexes.
+    group_ids_with_external_issue_ids = PlatformExternalIssue.objects.values_list(
+        "group_id", flat=True
+    )
```
Can someone fact-check me here 😬 ... might be nice to test this query against the production sentry.io database?

An alternative I was considering was to add:
```py
        # To avoid this subquery from returning data across all projects in the database
        group__project_id__in=[p.id for p in projects]
```
inside of a filter.

saifelse commented on src/sentry/search/snuba/backend.py:
```diff
@@ -49,6 +51,57 @@ def unassigned_filter(unassigned, projects):
+        relationship=GroupLink.Relationship.references,
+    ).values_list("group_id", flat=True)
+
+    # The naive way, implemented here, combines the two subexpressions with an OR.
+    # This may notably introduce suboptimal query plans due to two issues:
+    #   - There are no statistics on the overlap of the two subexpressions
+    #   - The planner might not realize it can use an optimized semi-join
+    # It's possible that these suboptimalities might not be the bottleneck for perf
+    # in production. But if they are, we may want to optimize for the happy path
```
@iProgramStuff : I unfortunately can't test this myself very easily since I don't have access to a representative database 😅 . Would it be possible to run some queries against the production sentry.io postgres database to test performance? 

For the following queries, I've used `project_id IN (111)`, but it would be good to repeat the following queries for 3 projects, where each fall into one of 3 categories respectively:
- Has many groups that are linked to ExternalIssue but none to PlatformExternalIssue.
- Has many groups that are linked to PlatformExternalIssue but none to ExternalIssue.
- Has groups that are linked to a mix of ExternalIssue and PlatformExternalIssue.

## OR query
### From PR
```sql
-- OR QUERY
-- from pr
EXPLAIN ANALYZE SELECT "sentry_groupedmessage"."id" FROM "sentry_groupedmessage"
WHERE (
    "sentry_groupedmessage"."project_id" IN (111)
    AND ("sentry_groupedmessage"."id" IN (
        SELECT U0."group_id" AS Col1 FROM "sentry_grouplink" U0 WHERE (
            U0."project_id" IN (111) AND U0."linked_type" = 3 AND U0."relationship" = 2
        )
        ) OR "sentry_groupedmessage"."id" IN (
            SELECT U0."group_id" AS Col1 FROM "sentry_platformexternalissue" U0
        )
    )
)
LIMIT 5001;
```
### Alternative query with group filter
```sql
SELECT "sentry_groupedmessage"."id" FROM "sentry_groupedmessage" WHERE (
    "sentry_groupedmessage"."project_id" IN (111)
    AND (
        "sentry_groupedmessage"."id" IN (
            SELECT U0."group_id" AS Col1
            FROM "sentry_grouplink" U0
            WHERE (U0."project_id" IN (111) AND U0."linked_type" = 3 AND U0."relationship" = 2)
        )
        OR "sentry_groupedmessage"."id" IN (
            SELECT V0."group_id" AS Col1
            FROM "sentry_platformexternalissue" V0
            WHERE V0."group_id" IN (
                SELECT U0."id" AS Col1
                FROM "sentry_groupedmessage" U0 WHERE U0."project_id" IN (111)
            )
        )
    )
) LIMIT 5001;
```
 
## PlatformExternalIssue happy path
### Integration query
```sql
EXPLAIN ANALYZE SELECT EXISTS(
   SELECT 1 FROM sentry_platformexternalissue
   JOIN sentry_groupedmessage ON
        sentry_groupedmessage.id = sentry_platformexternalissue.group_id
   WHERE sentry_groupedmessage.project_id IN (111)
)
```

### Issue query
```sql
SELECT "sentry_groupedmessage"."id" FROM "sentry_groupedmessage"
WHERE (
    "sentry_groupedmessage"."project_id" IN (111)
    AND "sentry_groupedmessage"."id" IN (
        SELECT U0."group_id" AS Col1 FROM "sentry_platformexternalissue" U0
    )
)
LIMIT 5001;
```

## ExternalIssue happy path
### Integration query
```sql
EXPLAIN ANALYZE SELECT EXISTS(
   SELECT 1 FROM sentry_grouplink
   WHERE sentry_grouplink.project_id IN (111)
)
```

### Issue query
```sql
EXPLAIN ANALYZE SELECT "sentry_groupedmessage"."id" FROM "sentry_groupedmessage"
WHERE (
    "sentry_groupedmessage"."project_id" IN (111)
    AND "sentry_groupedmessage"."id" IN (
        SELECT U0."group_id" AS Col1 FROM "sentry_grouplink" U0 WHERE (
            U0."project_id" IN (111) AND U0."linked_type" = 3 AND U0."relationship" = 2
        )
    )
)
LIMIT 5001;
```

BYK replied:
Maybe we can limit these filters behind a feature flag first to test it out?

saifelse replied:
Seems reasonable to me. I can add an OrganizationFeature called `organizations:is-linked-search` ?
Though, I think it'll still be helpful to run these `EXPLAIN ANALYZE`s to confirm the performance at various corner cases

saifelse replied:
@wedamija : An alternative code that I was alluding to is:

```py
platform_qs = PlatformExternalIssue.objects.filter(project_id__in=[p.id for p in projects])
integration_qs = GroupLink.objects.filter(
    project_id__in=[p.id for p in projects],
    linked_type=GroupLink.LinkedType.issue,
    relationship=GroupLink.Relationship.references,
)

group_linked_to_platform_issue_q = Q(id__in=platform_qs.values_list("group_id", flat=True))
group_linked_to_integration_issue_q = Q(id__in=integration_qs.values_list("group_id", flat=True))

# Usually a user will either only have PlatformExternalIssues or only have ExternalIssues,
# i.e. in most cases, at most one of the below expressions evaluates to True:
platform_issue_exists = platform_qs.exists()
integration_issue_exists = integration_qs.exists()
# By optimizing for this case, we're able to produce a filter that roughly translates to
# `WHERE group_id IN (SELECT group_id FROM one_issue_table WHERE ...)`, which the planner
# is able to optimize with the semi-join strategy.
if platform_issue_exists and not integration_issue_exists:
    query = group_linked_to_platform_issue_q
elif integration_issue_exists and not platform_issue_exists:
    query = group_linked_to_integration_issue_q
# ...but if we don't have exactly one type of issues, fallback to doing the OR.
else:
    query = group_linked_to_platform_issue_q | group_linked_to_integration_issue_q
```

Getting the times / plans for the following 4 queries would be useful to confirm if this alternative strategy is more performant.

```sql
-- Current query that does an OR
EXPLAIN ANALYZE 
SELECT "sentry_groupedmessage"."id" 
FROM "sentry_groupedmessage" WHERE (
    "sentry_groupedmessage"."project_id" IN (<redacted>)
    AND (
        "sentry_groupedmessage"."id" IN (
            SELECT U0."group_id" AS Col1
            FROM "sentry_grouplink" U0
            WHERE (U0."project_id" IN (<redacted>) AND U0."linked_type" = 3 AND U0."relationship" = 2)
        )
        OR "sentry_groupedmessage"."id" IN (
            SELECT V0."group_id" AS Col1
            FROM "sentry_platformexternalissue" V0
            WHERE (V0."project_id" IN (<redacted>) 
        )
    )
) LIMIT 5001; 
```

vs.

A
```sql
-- Existence of integration external issue
EXPLAIN ANALYZE SELECT EXISTS(
   SELECT 1 FROM sentry_grouplink
   WHERE sentry_grouplink.project_id IN (111) AND linked_type = 3 AND relationship = 2
)
```

B
```sql
-- Existence of platform external issue
EXPLAIN ANALYZE SELECT EXISTS(
   SELECT 1 FROM sentry_platformexternalissue
   WHERE sentry_platformexternalissue.project_id IN (111)
)
```

C
```
-- Happypath for integration external issue
EXPLAIN ANALYZE 
SELECT "sentry_groupedmessage"."id" 
FROM "sentry_groupedmessage" WHERE (
    "sentry_groupedmessage"."project_id" IN (111)
    "sentry_groupedmessage"."id" IN (
            SELECT U0."group_id" AS Col1
            FROM "sentry_grouplink" U0
            WHERE (U0."project_id" IN (111) AND U0."linked_type" = 3 AND U0."relationship" = 2)
    )
) LIMIT 5001; 
```
D
```
-- Happy path for platform external issue
EXPLAIN ANALYZE 
SELECT "sentry_groupedmessage"."id" 
FROM "sentry_groupedmessage" WHERE (
    "sentry_groupedmessage"."project_id" IN (111)
    "sentry_groupedmessage"."id" IN (
            SELECT V0."group_id" AS Col1
            FROM "sentry_platformexternalissue" V0
            WHERE (V0."project_id" IN (111))
    )
) LIMIT 5001; 
```

The alternative code I suggested I think will be sensible, if we see that `A + B + C < X` and `A + B + D < X`.

The non-happy-path case takes time `A+B+C+D` which I wouldn't be too concerned about if it's more than `X`, but guessing it's likely a neglibible difference.

wedamija replied:
@saifelse I think this is a good suggestion. With the OR in place we only filter `groupedmessage` via the project id, which ends up really slow. This plan is from a subsequent run, but initially this took 13 seconds:

X
```
EXPLAIN ANALYZE 
SELECT "sentry_groupedmessage"."id" 
FROM "sentry_groupedmessage" 
WHERE "sentry_groupedmessage"."project_id" IN (<redacted>)
AND (
    "sentry_groupedmessage"."id" IN (
        SELECT U0."group_id" AS Col1
        FROM "sentry_grouplink" U0
        WHERE (U0."project_id" IN (<redacted>) AND U0."linked_type" = 3 AND U0."relationship" = 2)
    )
    OR 
    "sentry_groupedmessage"."id" IN (
        SELECT V0."group_id" AS Col1
        FROM "sentry_platformexternalissue" V0
        WHERE (V0."project_id" IN (<redacted>))
    )
)
LIMIT 5001;

Limit (cost=994.88..2613.87 rows=5001 width=8) (actual time=4.004..1336.584 rows=426 loops=1)
-> Index Only Scan using sentry_groupedmessage_project_id_729c5589c650caf6 on sentry_groupedmessage (cost=994.88..419233.21 rows=1291921 width=8) (actual time=4.003..1336.416 rows=426 loops=1)
Index Cond: (project_id = <redacted>)
Filter: ((hashed SubPlan 1) OR (hashed SubPlan 2))
Rows Removed by Filter: 1836401
Heap Fetches: 266499
SubPlan 1
-> Bitmap Heap Scan on sentry_grouplink u0 (cost=10.00..586.61 rows=346 width=8) (actual time=0.152..2.576 rows=441 loops=1)
Recheck Cond: (project_id = <redacted>)
Filter: ((linked_type = 3) AND (relationship = 2))
Heap Blocks: exact=301
-> Bitmap Index Scan on sentry_grouplink_project_id (cost=0.00..9.91 rows=466 width=0) (actual time=0.094..0.094 rows=441 loops=1)
Index Cond: (project_id = <redacted>)
SubPlan 2
-> Bitmap Heap Scan on sentry_platformexternalissue v0 (cost=7.28..405.86 rows=386 width=8) (actual time=0.105..1.095 rows=422 loops=1)
Recheck Cond: (project_id = <redacted>)
Heap Blocks: exact=167
-> Bitmap Index Scan on sentry_platformexternalissue_project_id_ff4e3d10 (cost=0.00..7.18 rows=386 width=0) (actual time=0.080..0.080 rows=422 loops=1)
Index Cond: (project_id = <redacted>)
Planning time: 3.096 ms
Execution time: 1336.923 ms
```

VS doing just one statement at a time, we actually and up searching the index with `project_id` and `group_id`, much nicer:
C
```
EXPLAIN ANALYZE 
SELECT "sentry_groupedmessage"."id" 
FROM "sentry_groupedmessage" 
WHERE "sentry_groupedmessage"."project_id" IN (<redacted>)
AND (
    "sentry_groupedmessage"."id" IN (
        SELECT U0."group_id" AS Col1
        FROM "sentry_grouplink" U0
        WHERE (U0."project_id" IN (<redacted>) AND U0."linked_type" = 3 AND U0."relationship" = 2)
    )

)
LIMIT 5001;

Limit (cost=588.05..1564.54 rows=3 width=8) (actual time=3.359..18.947 rows=426 loops=1)
-> Nested Loop (cost=588.05..1564.54 rows=3 width=8) (actual time=3.356..18.880 rows=426 loops=1)
-> HashAggregate (cost=587.48..590.94 rows=346 width=8) (actual time=3.193..3.348 rows=426 loops=1)
Group Key: u0.group_id
-> Bitmap Heap Scan on sentry_grouplink u0 (cost=10.00..586.61 rows=346 width=8) (actual time=0.332..3.060 rows=441 loops=1)
Recheck Cond: (project_id = <redacted>)
Filter: ((linked_type = 3) AND (relationship = 2))
Heap Blocks: exact=301
-> Bitmap Index Scan on sentry_grouplink_project_id (cost=0.00..9.91 rows=466 width=0) (actual time=0.285..0.285 rows=441 loops=1)
Index Cond: (project_id = <redacted>)
-> Index Only Scan using sentry_groupedmessage_project_id_729c5589c650caf6 on sentry_groupedmessage (cost=0.57..2.80 rows=1 width=8) (actual time=0.026..0.036 rows=1 loops=426)
Index Cond: ((project_id = <redacted>) AND (id = u0.group_id))
Heap Fetches: 5895
Planning time: 7.304 ms
Execution time: 19.184 ms
```
and
D
```
EXPLAIN ANALYZE 
SELECT "sentry_groupedmessage"."id" 
FROM "sentry_groupedmessage" 
WHERE "sentry_groupedmessage"."project_id" IN (<redacted>)
AND (
    "sentry_groupedmessage"."id" IN (
        SELECT V0."group_id" AS Col1
        FROM "sentry_platformexternalissue" V0
        WHERE (V0."project_id" IN (<redacted>))
    )
)
LIMIT 5001;

Limit (cost=407.40..1496.29 rows=4 width=8) (actual time=1.448..12.650 rows=422 loops=1)
-> Nested Loop (cost=407.40..1496.29 rows=4 width=8) (actual time=1.447..12.599 rows=422 loops=1)
-> HashAggregate (cost=406.83..410.69 rows=386 width=8) (actual time=1.269..1.346 rows=422 loops=1)
Group Key: v0.group_id
-> Bitmap Heap Scan on sentry_platformexternalissue v0 (cost=7.28..405.86 rows=386 width=8) (actual time=0.151..1.182 rows=422 loops=1)
Recheck Cond: (project_id = <redacted>)
Heap Blocks: exact=167
-> Bitmap Index Scan on sentry_platformexternalissue_project_id_ff4e3d10 (cost=0.00..7.18 rows=386 width=0) (actual time=0.130..0.130 rows=422 loops=1)
Index Cond: (project_id = <redacted>)
-> Index Only Scan using sentry_groupedmessage_project_id_729c5589c650caf6 on sentry_groupedmessage (cost=0.57..2.80 rows=1 width=8) (actual time=0.019..0.026 rows=1 loops=422)
Index Cond: ((project_id = <redacted>) AND (id = v0.group_id))
Heap Fetches: 5953
Planning time: 4.029 ms
Execution time: 12.799 ms
```

Even with the extra IO wait time from the extra queries I think this is a better option. Just for completeness, here are the query plans for A and B as well.
A
```
EXPLAIN ANALYZE SELECT EXISTS(
   SELECT 1 FROM sentry_grouplink
   WHERE sentry_grouplink.project_id IN (<redacted>) AND linked_type = 3 AND relationship = 2
)

Result (cost=2.73..2.74 rows=1 width=1) (actual time=0.088..0.088 rows=1 loops=1)
InitPlan 1 (returns $0)
-> Index Scan using sentry_grouplink_project_id on sentry_grouplink (cost=0.42..799.11 rows=346 width=0) (actual time=0.087..0.087 rows=1 loops=1)
Index Cond: (project_id = 2277697)
Filter: ((linked_type = 3) AND (relationship = 2))
Planning time: 1.146 ms
Execution time: 0.121 ms
```

B
```
explain analyze SELECT EXISTS(
   SELECT 1 FROM sentry_platformexternalissue
   WHERE sentry_platformexternalissue.project_id IN (<redacted>)
)

Result (cost=0.92..0.93 rows=1 width=1) (actual time=8.910..8.911 rows=1 loops=1)
InitPlan 1 (returns $0)
-> Index Only Scan using sentry_platformexternalissue_project_id_ff4e3d10 on sentry_platformexternalissue (cost=0.29..243.02 rows=386 width=0) (actual time=8.906..8.906 rows=1 loops=1)
Index Cond: (project_id = 2277697)
Heap Fetches: 1
Planning time: 12.794 ms
Execution time: 9.816 ms
```

saifelse replied:
@wedamija : applied the changes!

saifelse commented on src/sentry/search/snuba/backend.py:
```diff
@@ -49,6 +51,57 @@ def unassigned_filter(unassigned, projects):
+    # scan over the Group rows for the given projects, which would negate the
+    # benefits of any indexes used for other filters. This would be more reasonable if
+    # we had a redundant `project_id` (like GroupLink.project_id) on this table.
+    group_ids_with_platform_external_issue = PlatformExternalIssue.objects.values_list(
```
@wedamija : thanks! Happy to poke around adding a migration.

In my mind the correct order of steps is:
1. Add `project_id = BoundedBigIntegerField(null=True, db_index=True)` to `PlatformExternalIssue`, and run in production. I'm assuming like GroupLink, this won't be an FK.
2. Update backend to start setting `project_id` column for new `PlatformExternalIssue`s.
3. Run a data migration that backfills `project_id` which then make project_id non-nullable. Since it only has 10K rows, this can be done in a single transaction.
4. Land this PR now that `project_id` is correctly maintained.

Unless there's a sophisticated system for migrations and the best we have is all migrations are run before deploy, it looks like I can break this down into at least two PRs:
- One PR with 1&2 (or split into two PRs)
- After 1&2 are deployed, a PR that does 3&4 can land / deploys (or split into two PRs).

Does that sound right?

saifelse replied:
1+2: https://github.com/getsentry/sentry/pull/21858

wedamija commented:
/gcbrun

wedamija reviewed (approved):
This is an epic feature you've built here, thanks for doing this! I really appreciate the thorough tests as well.

I'll merge this soon, although it looks like ci is failing for unrelated reasons. Could you rebase this? That should resolve the issues.

BYK commented:
/gcbrun

EvanPurkhiser commented:
wow this is a great feature. Thank you for building this!

chadwhitacre commented:
Heck yeah, thanks for the contribution @saifelse! 💃

saifelse commented:
> This is an epic feature you've built here, thanks for doing this! I really appreciate the thorough tests as well.

> wow this is a great feature. Thank you for building this!

> Heck yeah, thanks for the contribution @saifelse! 💃

Thanks everyone! Expect to see more issues and PRs from me around issue assignment / ownership and the Slack integration 🙃
